DRF Expansion

rneswold opened the pull request:
This pull request is offered up for discussion -- it's not ready to be merged, if at all.

rneswold committed:
:memo: document some messages

rneswold committed:
:construction: enhance DRF messages

Add arm, sample, and stop events to the specification.

rneswold committed:
:bug: field names need to be different

Cut-n-paste bug. :/

rneswold committed:
:recycle: add time to arm and stop events

beauremus commented:
@KitFieldhouse, I'm interested in whether these structures match your arm and trigger functionality.

beauremus commented:
I got the help of an LLM to write some pseudo-code user examples to see how this "feels" from a user perspective.

```python
# Example 1: Requesting data periodically
# This example shows how to request a specific data channel ('device.name')
# with its primary field, sampled every 5 seconds, starting immediately.
request = DRF(
    data_channel=DataChannel(
        name="device.name",
        range=Range(full_range=google.protobuf.Empty()), # Is this required? Can it be left undefined?
        field=Field(reading=Field.Flavor.PRIMARY) # I think because `RAW` is first in the ENUM, it's the default
    ),
    sample=SampleEvent(
        periodic=SampleEvent.Periodic(
            period=google.protobuf.Duration(seconds=5),
            immediate=True
        )
    )
)

# Example 2: Requesting data on a TCLK accelerator event
# This example demonstrates requesting the 'device.name' raw data
# on TCLK event $1D.
request = DRF(
    data_channel=DataChannel(
        name="device.name",
        range=Range(full_range=google.protobuf.Empty()),
        field=Field(reading=Field.Flavor.RAW) # Is this required? Can it be left undefined?
    ),
    sample=SampleEvent(
        clock=common.event.Clock(
            event_number=123, # It's not a huge burden to have users convert from hex to decimal, but it's not great
            # Python would look something like this, I think
            event_number=int("1D", 16),
            event_type=common.event.Clock.EventType.HARDWARE # Should we consider deprecation?
        )
    )
)

# Example 3: Requesting data on a state event (value change)
# This example shows how to request data when 'state.monitor' changes its value.
request = DRF(
    data_channel=DataChannel(
        name="device.name",
        range=Range(full_range=google.protobuf.Empty()),
        field=Field(reading=Field.Flavor.COMMON)
    ),
    sample=SampleEvent(
        state=common.event.State(
            state_name="state.monitor",
            any_value=google.protobuf.Empty() # Trigger on any value change
        )
    )
)

# Example 4: Requesting data with a delay after a clock event
# This example requests 'device.name' data on TCLK event $FA,
# but with a 100 millisecond delay after the event fires.
request = DRF(
    data_channel=DataChannel(
        name="device.name",
        range=Range(full_range=google.protobuf.Empty()),
        # Note that this test shows an error state because `setting` doesn't have a `NOMINAL` field
        # It would be a lot of duplication but should we consider an ENUM per field?
        field=Field(setting=Field.Flavor.NOMINAL)
    ),
    sample=SampleEvent(
        clock=common.event.Clock(
            event_number=int("FA", 16),
            event_type=common.event.Clock.EventType.HARDWARE,
            # second and nanos are the available arguments
            delay=google.protobuf.Duration(nanos=100_000_000) # 100 milliseconds
        )
    )
)

# Example 5: Requesting a specific array range periodically
# This example shows requesting elements from index 10 to 20 of an array device,
# sampled every 2 seconds.
request = DRF(
    data_channel=DataChannel(
        name="array.device",
        range=Range(
            array_range=ArrayRange(
                start_index=10,
                end_index=20
            )
        ),
        field=Field(reading=Field.Flavor.RAW)
    ),
    sample=SampleEvent(
        periodic=SampleEvent.Periodic(
            period=google.protobuf.Duration(seconds=2),
            immediate=True
        )
    )
)

# Example 6: Requesting control field to send a command (one-time)
# This example illustrates sending a control command to 'device.command'
# immediately, without continuous sampling.

# This API is reading-oriented and not setting-oriented
# This request doesn't make sense, but is valid
request = DRF(
    data_channel=DataChannel(
        name="device.command",
        range=Range(full_range=google.protobuf.Empty()),
        field=Field(control=google.protobuf.Empty())
    ),
    sample=SampleEvent(
        immediate=google.protobuf.Empty()
    )
)

# Does it make sense if we remove the sample?
request = DRF(
    data_channel=DataChannel(
        name="device.command",
        range=Range(full_range=google.protobuf.Empty()),
        field=Field(control=google.protobuf.Empty())
    )
)

# Modifying the DAQ service to accept DRF instead of a string
DAQ.Set([
    device=request, # See above
    value=Value(boolean=True) # We do not currently support `Control` actions in `common.device.Value`
])

# Example 7: Arming a data acquisition to stop after a certain count
# This example shows how to arm a data acquisition for 'device.status'
# to start on a hardware clock event $0C, and stop after 5 samples.
request = DRF(
    data_channel=DataChannel(
        name="device.status",
        range=Range(full_range=google.protobuf.Empty()),
        field=Field(status=Field.Flavor.ON)
    ),
    arm=ArmEvent(
        clock=common.event.Clock(
            event_number=int("0C", 16),
            event_type=common.event.Clock.EventType.HARDWARE
        )
    ),
    stop=StopEvent(
        count=5
    )
)

# Example 8: Requesting multiple data channels periodically
# This example shows how to request primary field data from 'device.one'
# and raw data from 'device.two' simultaneously, both sampled every 10 seconds.
request = DRFs(
    data_channel=[
        DataChannel(
            name="device.one",
            range=Range(full_range=google.protobuf.Empty()),
            field=Field(reading=Field.Flavor.PRIMARY)
        ),
        DataChannel(
            name="device.two",
            range=Range(full_range=google.protobuf.Empty()),
            field=Field(reading=Field.Flavor.RAW)
        )
    ],
    sample=SampleEvent(
        periodic=SampleEvent.Periodic(
            period=google.protobuf.Duration(seconds=10),
            immediate=True
        )
    )
)
```

beauremus commented on proto/controls/common/v1/drf.proto:
```diff
@@ -25,14 +31,18 @@ message DataChannel {
+    common.event.SampleEvent sample = 3;
+    common.event.StopEvent stop = 4;
+    optional common.sources.Source source = 5;
 }
```
Does this request "feel" good to you?

> sample: Periodic 1440Hz
> arm: 0x12
> trigger: 0x1D
> stop: 0X1F

```python
request = DRF(
    data_channel=DataChannel(
        name="device.status",
        range=Range(full_range=google.protobuf.Empty()),
        field=Field(status=Field.Flavor.ON)
    ),
    sample=SampleEvent(
        periodic=SampleEvent.Periodic(
            # This stinks and could be an issue with how this is "rounded"
            period=google.protobuf.Duration(nanos=math.trunc(1/1440*1_000_000_000)),
            immediate=True
        )
    ),
    arm=ArmEvent(
        clock=common.event.Clock(
            event_number=int("12", 16),
            event_type=common.event.Clock.EventType.HARDWARE
        )
    ),
    trigger=TriggerEvent(
        clock=common.event.Clock(
            event_number=int("1D", 16),
            event_type=common.event.Clock.EventType.HARDWARE
        )
    ),
    stop=StopEvent(
        clock=common.event.Clock(
            event_number=int("1F", 16),
            event_type=common.event.Clock.EventType.HARDWARE
        )
    )
)
```

> sample: 1440Hz
> arm: 0x1D
> trigger: 0x52
> stop: 0x1F

```python
request = DRF(
    data_channel=DataChannel(
        name="device.status",
        range=Range(full_range=google.protobuf.Empty()),
        field=Field(status=Field.Flavor.ON)
    ),
    sample=SampleEvent(
        periodic=SampleEvent.Periodic(
            # This stinks and could be an issue with how this is "rounded"
            period=google.protobuf.Duration(nanos=math.trunc(1/1440*1_000_000_000)),
            immediate=True
        )
    ),
    arm=ArmEvent(
        clock=common.event.Clock(
            event_number=int("1D", 16),
            event_type=common.event.Clock.EventType.HARDWARE
        )
    ),
    trigger=TriggerEvent(
        clock=common.event.Clock(
            event_number=int("52", 16),
            event_type=common.event.Clock.EventType.HARDWARE
        )
    ),
    stop=StopEvent(
        clock=common.event.Clock(
            event_number=int("1F", 16),
            event_type=common.event.Clock.EventType.HARDWARE
        )
    )
)
```

I think this would require something like:
```proto
message TriggerEvent {
    oneof event {
        google.protobuf.Empty immediate = 1; # This is the default, so a no-op
	Clock clock = 2;
	State state = 3;
	google.protobuf.Duration delay = 4;
    }
}
```

beauremus replied:
Speaking to "complex trigger conditions" I was considering posting this and decided not to. Does this align with what you were thinking?

```proto
// Represents a single condition within a compound event (either a Clock or a State).
message EventCondition {
    oneof condition_type {
        Clock clock = 1;
        State state = 2;
    }
}

// Defines a compound event by combining multiple EventConditions with logical operators.
message CompoundEvent {
    enum Operator {
        AND = 0;      // All conditions must be met (e.g., all clocks fire near-simultaneously)
        OR = 1;       // Any one condition must be met
        SEQUENCE = 2; // Conditions must be met in the specified order
    }

    Operator operator = 1;
    repeated EventCondition conditions = 2; // List of individual event conditions
}
```

KitFieldhouse replied:
Not sure if this is terribly useful, but I had done some thinking about complex trigger conditions for the MI studies logger project and came up with some ideas for different trigger types (some of which are already in this proto):

- Periodic Trigger: A trigger that is periodically satisfied at some interval.
- DateTime Trigger: A trigger that is satisfied only once at a given absolute time
- TCLK Trigger: Satisfied when the TCLK event goes out
- Data Trigger: Satisfied when some DRF returns a value in a range (I had in mind that this could be a different drf than what we are applying the trigger on, for instance, I could say give me R:LI110 but only when R:BEAM is above 1e12 but less than 3e12).
- User Trigger: This would require a client to manually satisfy the trigger through some API call
- Interval Trigger: This trigger would satisfy a fixed amount of time from some other trigger

With these you could then build up a trigger sequence and then specify within that trigger sequence when you want data to be returned. For instance, to grab the I:BEAM at 1440HZ for the first 2A with beam every hour:

The sequence:
Interval trigger every hour -> TCLK 52 ->  TCLK 2A -> TCLK 26

Sample: 
I:BEAM at 1440HZ from TCLK 52 to TCLK 26

The problem with complicated triggering conditions like this is that I think you end up having to buffer data as you wait for trigger conditions to be met. In the above example, we do not know as soon as we receive a TCLK 52 that it is going to be followed by a TCLK 2A so you have to end up buffering data until the rest of the sequence is satisfied.

In my mind, I think we should keep the DRF triggering conditions relatively simple, supporting only triggering conditions that do not require complicated buffering of data and instead can operate directly on the data stream. For more complicated triggering conditions, I think this should be a separate API to a dedicated service.

rneswold replied:
> - Data trigger: Satisfied when ...

I don't like the idea of using device names in the specs because now you have something that might not exist and you have to report an error. I believe all our messages, up to this point, generate "correct" specifications (even if it specifies a clock event that doesn't fire, for example) and so never have to report an error. Using device names changes that.

> - User Trigger: This would require ...

User triggers introduce a latency. I would simply set a state device as a user input channel.

beauremus replied:
> I guess I need to know how a "user trigger" works. Is it a hardware signal that a user can activate? Or a gRPC request? Setting a state device would be similar complexity as the latter.

I had assumed the latter. @KitFieldhouse did you mean something different?

beauremus commented on proto/controls/common/v1/event.proto:
```diff
@@ -3,48 +3,79 @@ syntax = "proto3";
+
+    oneof event {
+        google.protobuf.Empty never = 1;
+        google.protobuf.Empty immediate = 2;
```
We use `never` in DRF like this when users want to make a setting.
Maybe that use case doesn't make sense, because you should request a response to validate the setting?

I think my setting example shows that we don't need this 🤔 

```python
# Does it make sense if we remove the sample?
request = DRF(
    data_channel=DataChannel(
        name="device.command",
        range=Range(full_range=google.protobuf.Empty()),
        field=Field(control=google.protobuf.Empty())
    )
)

# Modifying the DAQ service to accept DRF instead of a string
DAQ.Set([
    device=request, # See above
    value=Value(boolean=True) # We do not currently support `Control` actions in `common.device.Value`
])
```

rneswold replied:
I made a change so that `never` is part of the arming event.

In this protocol, making a setting only indicates success of failure -- you don't get a reading (or stream of readings.) The setting API only takes a device name and a setting value.

I'm resolving this conversation since it's referring to field that's been moved.

rneswold committed:
:recycle: Adjust how to support post trigger

This reorganization is a bit simpler than the previous and still allows
post triggers. Rather than having an arm event indicating the arming
goes at the end of collection, we just use the stop event. But that
means we can't use the `count` field, so that field was moved to the
`SampleEvent` message as a separate field. If 0, it means "infinite"
samples, otherwise it indicates an upper limit of how many points will
be sent.

Any acquisition service should treat a `StopEvent` that isn't `never`
and a `SampleEvent` with a `count` not zero as a post-trigger request;
that is, it keeps a circular buffer of the most recent `count` readings
and sends them when the `StopEvent` occurs.

rneswold committed:
:bug: move `never` to `ArmEvent`

rneswold commented:
With 5cf99ab, if you want to see the data (1,000 points) before an event ($21) happened, @1440Hz, you'd do:

```
device : R:IBEAM
arm: immediate
sample: periodic @ 1440Hz, count: 1,000
stop: clock $21
```

rneswold commented on proto/controls/common/v1/event.proto:
```diff
@@ -3,48 +3,79 @@ syntax = "proto3";
+message SampleEvent {
+    message Periodic {
+        google.protobuf.Duration period = 1;
+        bool immediate = 2;
```
`no_delay`?

rneswold commented on proto/controls/common/v1/drf.proto:
```diff
@@ -25,14 +31,18 @@ message DataChannel {
     DataChannel data_channel = 1;
-    optional common.event.Event event = 2;
-    optional common.sources.Source source = 3;
+    common.event.ArmEvent event = 2;
```
Are we proposing to combine them in another message?

beauremus replied:
I think that's best. It adds another layer, but it'll be less error-prone.
